The initial commit of selenium environment

mutirri opened the pull request:
First draft of selenium environment for bokeh.
Please read `README.md` file and follow it recommendations.

damianavila commented:
>This is friendly reminder that, according to [BEP-1](https://github.com/bokeh/bokeh/wiki/BEP-1:-Issues-and-PRs-management#pull-requests) PRs management guidelines, the current PR will be closed soon if no further activity is detected. You can keep working on/pushing to your branch and reopen the PR when you believe is ready for feedback again. 

Now the question? @mutirri what's are your plans on this PR? there still seems to be some problems with py3 yet...

mutirri commented:
Hi Damian,

> Now the question? @mutirri what's are your plans on this PR?

Right now I'm waiting for some feedback from Bryan.
@bryevdv did you have a chance to look on this pull request?
The selenium code responsible of performing the tests works ok.
However initialization of bokeh server is not implemented properly - please look at the `create_app` function in `RawSeleniumTestFixture` class (`fixtures.py` file).

> there still seems to be some problems with py3 yet...

I have adjusted the code according to previous suggestions.
Exactly what problems do you mean?

damianavila commented:
>Exactly what problems do you mean?

TravisCI is failing because of this:

`Hint: flask-testing has a conflict with the remaining packages`

mutirri commented:
Hi Damian,

Yeah, you were right. I don't know how I missed it.
Thanks for pointing me to this.
Now all packages are available directly on bokeh organization (flask-testing, blinker) - for python 3.3 as well as for python 3.4.

bryevdv commented on bokeh/tests/selenium/config.py:
```diff
@@ -0,0 +1,215 @@
+        if 'data_dir' in kwargs:
+            self.data_dir = kwargs.get('data_dir')
+        else:
+            self.data_dir = None
```
All of these `if...else` blocks should be expressible as just `kwargs.get('data_dir', None)`

mutirri replied:
Done.

bryevdv commented on bokeh/tests/selenium/README.md:
````diff
@@ -0,0 +1,180 @@
+```
+$ conda install selenium-server-standalone
+$ conda install selenium-chromedriver
+$ conda install phantomjs
````
no osx conda pkg for this

mutirri replied:
Now almost all packages are available for osx-64 platform also.
Still I don't have an access to the raw MacOSX machine.
Adiditionally on binstar.org are no build workers for osx-64 platform and `conda convert` subcommand  is not working for `phantomjs` package (it's an ELF executable and it can't be converted in this way).

bryevdv commented on bokeh/tests/selenium/README.md:
````diff
@@ -0,0 +1,180 @@
+
+```
+$ conda install selenium-server-standalone
+$ conda install selenium-chromedriver
````
since instructions are different for OS X for this pkg, move it below it one "if you are not on OS X do this, other wise do that" block. I almost ran this command before seeing the note about OS X below

mutirri replied:
Done.

bryevdv commented on bokeh/tests/selenium/README.md:
````diff
@@ -0,0 +1,180 @@
+1. To get general help, try:
+
+```
+$ ./run_selenium_tests.py -h
````
what directory is this in?

mutirri replied:
Uhm, the script is in the directory where README.md file is placed.
Anyway additional info was added to documentation.

bryevdv commented on conda.recipe/meta.yaml:
```diff
@@ -57,6 +57,11 @@ test:
     - websocket-client
     - ipython
     - ipython-notebook
+    - flask-testing
```
no OSX conda pkg

mutirri replied:
Added `flask-testing` packages for osx-64 platform (py27, py33, py34).

bryevdv commented:
Hi @mutirri since I don't know very much at all about the details of Selenium, we will have to rely very heavily on your leadership for this. I have made a few comments and also noted some pkgs that are missing for OS X. Once I eventually got everything installed, I was able to get this:
```
Discovering tests in /Users/bryan/work/bokeh/bokeh/tests/selenium/tests ...
 * Running on http://127.0.0.1:5000/
.
----------------------------------------------------------------------
Ran 1 test in 4.006s

OK
```
So that seems like a good start? I would say please proceed in whatever way you feel most appropriate as far as the implementation goes. I think the next best high level goal is to add two or three actual simple tests so we can see what it looks like to specify them. Then lets see about enabling just these couple of tests on TravisCI. Ultimately I would like a fairy extensive Selenium test suite but it that will probably need to be in its own separate test run, hopefully on Binstar, but still waiting on that integration.

bryevdv commented:
I should add that our current TravisCI times seem to be around ~25 minutes, so adding two or three selenium tests should hopefully be possible. And even just a couple of tests would already provide a big marginal improvement in confidence, since it is often the case that if any one server/UI example is broken, they all are. Not always of course but it's still a very good start.

damianavila commented:
@mutirri what are your next steps on this one? Do you plan to add more things here? Iterate with other PRs? Thanks.

mutirri commented:
Hi,

> since I don't know very much at all about the details of Selenium,
> we will have to rely very heavily on your leadership for this.

Of course. However current problem is not related to Selenium code.
Problem is just related to the code responsible of bokeh initialization
(exactly: implementation of `bokeh_app` and its use by
`LiveServerTestCase` - http://flask-testing.readthedocs.org/en/latest/).

I) Normal bokeh-server startup:
```
$ bokeh-server
```
switch to another console
```
$ ipython
>>> import urllib2
>>> ret = urllib2.urlopen('http://localhost:5006')
>>> ret.code
200
```

So, everything is OK.

II) Running `bokeh-server` (flask application) in testing mode (http://flask-testing.readthedocs.org/en/latest/):

```
$ python scripts/runtests.py
```

What is happening now (short version limited only to selenium test):

a) The initialization of `test_sample` is started - running all stuff from fixtures on which this test depends on, so:
b) Running the `BasicSeleniumTestFixture` which depends on `RawSeleniumTestFixture`.
c) Running the `RawSeleniumTestFixture` which depends on `LiveServerTestCase`.
This fixture contains reference to bokeh-server initialization code (exactly in `make_app` function, which function is required by `LiveServerTestCase`, to return whole bokeh application - with appropriately defined `run` method). So, I'm returning bokeh-server application (first `from bokeh.server.app import app as bokeh_app` and then returning `bokeh_app` in `make_app` as it was mentioned).
d) Running the `LiveServerTestCase` as the last dependency (this fixture starts bokeh-server through  the `bokeh_app` from previous fixture - by invoking its `run` method).
And now:
e) In `RawSeleniumFixture` is starting `check_if_server_is_up_and_running` function which returns 404 HTTP code.
Additionally it doesn't matter if I will connect in this function to some addresses that should be available under bokeh-server, like:
- http://127.0.0.1:5000/

or

- http://127.0.0.1:5000/bokeh/ping

because I'm always getting 404.

All of this means that address `http://127.0.0.1:5000` is available (`run` method was executed and `netstat` command shows appropriately entry) but server is somehow incompatibly initialized (according to `LiveServerTestCase`). **That's the problem**.

Additionally please check `bokeh/server/tests/test_utils.py` and `BaseBokehServerTestCase` also (it doesn't work for me).

> I have made a few comments and also noted some pkgs that are missing for OS X.

I adjusted code according to your suggestions.
I also added missing packages.
However keep in mind that there is still no `phantomjs` package for osx-64 (https://github.com/bokeh/bokeh/pull/1520#discussion_r22642766).

> Once I eventually got everything installed, I was able to get this:
> Ran 1 test in 4.006s
>
> OK
> So that seems like a good start?

Yes. This result means that selenium code is working OK (we are able to test applications).
However the current test checks the google site instead of bokeh-server due to its initialization problem.

> I would say please proceed in whatever way you feel most appropriate as far as the
> implementation goes.

Yeah, I'm trying to resolve this initialization problem unfortunately without luck so far.
It seems that I don't have enough knowledge of bokeh-server architecture.

> I think the next best high level goal is to add two or three actual simple tests so we can
> see what it looks like to specify them. Then lets see about enabling just these couple of tests
> on TravisCI. Ultimately I would like a fairy extensive Selenium test suite but it that will probably
> need to be in its own separate test run, hopefully on Binstar, but still waiting on that integration.

I agree.

> I should add that our current TravisCI times seem to be around ~25 minutes,
> so adding two or three selenium tests should hopefully be possible.
> And even just a couple of tests would already provide a big marginal improvement in confidence,
> since it is often the case that if any one server/UI example is broken, they all are.
> Not always of course but it's still a very good start.

Indeed.

bryevdv commented:
@mutirri Sorry for late reply. I have seen some cross-site issues prevent `127.0.0.1` from working, **but `localhost` does work**. Can you try with `localhost` instead?

mutirri commented:
Hi,

> @mutirri Sorry for late reply.

No problem. I know that you are busy.
In meanwhile I was doing my other tasks.

> I have seen some cross-site issues prevent 127.0.0.1 from working, but localhost does work.
> Can you try with localhost instead?

I tried this before as well. However under `localhost` the server is unavailable:
```
Test dir: tests
Test dir: tests/selenium/tests
Test dir: server/tests
Test dir: server/templates/tests

   Bokeh version: 0.7.0-53-g88ac008-dirty
  Python version: 2.7.8-CPython
        Platform: Linux-2.6.32-504.3.3.el6.x86_64-x86_64-with-centos-6.6-Final
    
E
======================================================================
ERROR: Simple test case.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/irritum/stuff/repos/my/bokeh/bokeh/tests/selenium/fixtures.py", line 79, in setUp
    RawSeleniumTestFixture.setUp(self)
  File "/home/irritum/stuff/repos/my/bokeh/bokeh/tests/selenium/fixtures.py", line 24, in setUp
    self.check_if_server_is_up_and_running()
  File "/home/irritum/stuff/repos/my/bokeh/bokeh/tests/selenium/fixtures.py", line 61, in check_if_server_is_up_and_running
    response = urlopen('http://localhost:5006/')
  File "/home/irritum/anaconda/envs/devbokeh/lib/python2.7/urllib2.py", line 127, in urlopen
    return _opener.open(url, data, timeout)
  File "/home/irritum/anaconda/envs/devbokeh/lib/python2.7/urllib2.py", line 404, in open
    response = self._open(req, data)
  File "/home/irritum/anaconda/envs/devbokeh/lib/python2.7/urllib2.py", line 422, in _open
    '_open', req)
  File "/home/irritum/anaconda/envs/devbokeh/lib/python2.7/urllib2.py", line 382, in _call_chain
    result = func(*args)
  File "/home/irritum/anaconda/envs/devbokeh/lib/python2.7/urllib2.py", line 1214, in http_open
    return self.do_open(httplib.HTTPConnection, req)
  File "/home/irritum/anaconda/envs/devbokeh/lib/python2.7/urllib2.py", line 1184, in do_open
    raise URLError(err)
URLError: <urlopen error [Errno 111] Connection refused>

----------------------------------------------------------------------
Ran 1 test in 1.354s

FAILED (errors=1)
```

So, in summary:
a) by using 127.0.0.1 (this address is returned from original `get_server_url` function from `LiveServerTestCase`) - server seems to be running but it is responding with 404 (auth problem?).
b) by using localhost (manual change of address in code) - server is not responding at all.

In general the code responsible of bokeh server initialization is quite unusual/complicated (detached `run` function from `app` flask application, start process is splitted into couple of minor stages, alternate initialization in test files - by using threads module - in compared to default server start up which uses the `tornado` HTTP server).

bryevdv commented:
@mutirri One option: bokeh-server should expose a Flask Blueprint. You should be able to create a new app that incorporates the blueprint.In fact, getting (and documenting) experience with deploying bokeh-server in different ways would be extremely valuable. So don't feel like you have to be tied to the specific `bokeh-server` script. Does this seem like a potential way forward?

damianavila commented:
I could not found an issue associated, so I will label this as feature... then, we have to milestone it. I don't think this would be ready before 0.8... thoughts? @bryevdv @mattpap

bryevdv commented:
I have not had a chance to look at it recently. @mutirri any updates?

mutirri commented:
> @mutirri any updates?
@bryevdv no updates. I dealt with other tasks.
I only read Hugo's email which was send to dev list and descibing briefly many levels of server initialization and one of existing issue - related to gunicorn and shutting down of server.

mutirri commented:
Hi @bryevdv,

1. Is there any possibility to load for example `test.py` file (with code of bokeh plot) into running server through GUI? For example I would like to click on `add` button on main page and then insert path to my document.

2. I'm aware of `output_server()` function, however it always execute a new tab in my browser and I don't want it. Can I turn off this behavior somehow?

3. Maybe there is another functionality responsible of loading files directly to running bokeh-server?

bryevdv commented:
@mutirri It's the `show(...)` calls that raise a browser window. We can write examples that don't do that, or you can monkey patch `show()` to just do a `push()`

I'm not sure about the first one, the server is loaded by running a python process, so it seems the button would have to run a python process.

mutirri commented:
Hi @bryevdv,

Please add a new commit with CSS handler for each tool in toolbar.
That's because I need this to properly pick elements from web (unique handler of each element).
For example, inactive `Box Select` tool has this CSS applied:
```
<button type='button' class='bk-toolbar-button hover'>
```
and I would like to have on it:
```
<button type='button' class='handler-select bk-toolbar-button hover'>
```

Of course `handler-select` CSS class should not exists (no body for this class is required).

bryevdv commented:
If `bk-handler-select` is good, I would prefer that (with the `bk-` prefix) but if it needs to be `handler-select` because that is what selenium is hard coded to expect that is OK. 

The CSS is applied in this `eco` template:

https://github.com/bokeh/bokeh/blob/master/bokehjs/src/coffee/tool/button_tool_template.eco

Can you make the change yourself, as part of this PR?

mattpap commented:
All bokeh's CSS classes have to have a prefix.

bryevdv commented:
If selenium dictates some specific CSS class name, it seems to me we would have no choice but to use that. I'm not saying it does, just explicitly mentioning the possibility. Of course in that case I would not really consider it a "bokeh" CSS class anyway, it would be a "selenium" CSS class.

mutirri commented:
Hi,

> If bk-handler-select is good, I would prefer that.

Sure. No problem.

> If selenium dictates some specific CSS class name, it seems to me we
> would have no choice but to use that.

It's free to use any name of CSS class.
I just need an unique name per element.

bryevdv commented:
Great, let's use `bk-handler-select`

birdsarah commented:
I haven't had a chance to read all the PR yet, but I wanted to chip in with a quick comment, and I'm happy to expand more if it's useful.

In another task I've seen, there's talk of moving to pytest. I wanted to highlight the pytest plugin [pytest-mozweqa](https://github.com/mozilla/pytest-mozwebqa). This plugin provides a whole host of useful boiler plate for working with selenium and is completely generic in spite of the name. In addition it: has hooks for working with SauceLabs (which would be free to bokeh as an open source project), and has nice reporting features which takes a screenshot at the point of failure and sticks it in an html report (e.g. http://birdsarah.github.io/pytest-mozwebqa-report-proto/issue_74_original_formatting/ - it may not be exactly like this styling anymore, but something similar)

birdsarah commented:
On a different topic. It's a bit heavy for bokeh's needs, but just checking we're explicitly choosing not to use the PageObject pattern.

mutirri commented:
Hi @bryevdv.

> Great, let's use `bk-handler-select`

Uhm, I added necessary handlers to all tools in this commit: https://github.com/mutirri/bokeh/commit/1f200607f307e7db2e1edd419490b86070f8e4ea
I tried to do this in the same manner as `@icon` is done.
However, each of tool has now exactly the same additionall CSS class - `bk-handler-box-select`,
even if theirs prototypes are different.
Any hints?

mutirri commented:
Hi @bryevdv,
Sorry for delay but I had to switch to some urgent internal task.
I added first interactive test which checks if plot was moved by user.
However to be able to add more of interactive tests I need some help with handlers (https://github.com/bokeh/bokeh/pull/1520#issuecomment-77157463).
Could you provide me some extra info how should I add them to each tool?

bryevdv commented:
@mutirri Sure, I just need to know more about exactly what you need to do. Do you need to add a specific, different CSS class to each toolbar button? If so, the button template is here:

https://github.com/bokeh/bokeh/blob/master/bokehjs/src/coffee/tool/button_tool_template.eco

I think you can just template in the `selenium_handler` property that you have added to the tool models and then it will be added to the the buttons when the template is instantiated, here:

https://github.com/bokeh/bokeh/blob/master/bokehjs/src/coffee/tool/button_tool.coffee#L20

Is that the question, or have I misunderstood?

bryevdv commented:
Also, thanks this will be a great addition to our testing machinery. I think maybe we could use this as a practice case for setting up some of our tests on binstar?

mutirri commented:
Hi @bryevdv,

> Do you need to add a **specific, different** CSS class to each toolbar button?

Exactly.

> If so, the button template is here:
> https://github.com/bokeh/bokeh/blob/master/bokehjs/src/coffee/tool/button_tool_template.eco

Yes I know that. I have even modified this template here: https://github.com/mutirri/bokeh/commit/1f200607f307e7db2e1edd419490b86070f8e4ea
Please look at the content of the first two diffs.

> I think you can just template in the selenium_handler property that you have added to the tool
> models and then it will be added to the the buttons when the template is instantiated, here:
> https://github.com/bokeh/bokeh/blob/master/bokehjs/src/coffee/tool/button_tool.coffee#L20

And this is exactly what I have already done. After compilation I can see my new CSS class on each tool button of toolbar **however** instead of getting **dedicated** CSS class **per tool** - I'm always getting **the same** CSS class (the first one by alphabetical order): `bk-handler-box-select`.

This means that for example `lasso` button gets just mentioned `bk-handler-box-select` CSS class instead of `bk-handler-lasso-select`.

That's the problem :)

> Also, thanks this will be a great addition to our testing machinery.

No problem.

> I think maybe we could use this as a practice case for setting up some of our tests on binstar?

Yeah. I will add more interactive tests for all of remaining tools (when I will be able to easily click on each of these tools). After that I will work on setting up tests to automatically launch the tests on binstar.

bryevdv commented:
@mutirri I am not sure, looking at the code, everything looks fine. Can you put a debugger breakpoint in the `initialize` that renders the template and inspect the model there? Alternatively, the other template parameter comes from a property, like this

https://github.com/mutirri/bokeh/blob/1f200607f307e7db2e1edd419490b86070f8e4ea/bokehjs/src/coffee/tool/button_tool.coffee#L39

 Maybe there is some weird interaction and it only works to fill in template values from properties. Worth a try tho if it works I'd like to understand. I will try to find some time to actually run and debut this today if I can.

bryevdv commented:
@birdsarah I don't know enough about selenium to comment on PageObject.

bryevdv commented:
@mutirri did you have any luck with a debugger?

mutirri commented:
Hi @bryevdv,

> @mutirri did you have any luck with a debugger?

Sorry but no.
I was forced fix some issues related to unexpected change of bucket names on amazon.
Recently I had a few days off also.
I'll look at this in this week.

mutirri commented:
Hi @bryevdv,

I'm having some troubles with debugging.
Right now I'm trying to setup my development environment once again because:
- I was forced to clean up some packages through `conda clean` subcommand - to gain more free space after 2 years of usage - this breaks a little bit my `devbokeh` environment,
- recently you have switched from `grunt` to `gulp` and made even more changes in code and overall development process (like: https://github.com/bokeh/bokeh/pull/2109),
- documentation on page http://bokeh.pydata.org/en/latest/docs/dev_guide/building.html#building-bokehjs is obsolate.

Currently I'm getting:
```
$ [irritum@localhost bokehjs]$ gulp

module.js:340
    throw err;
          ^
Error: Cannot find module 'coffee-script/register'
    at Function.Module._resolveFilename (module.js:338:15)
    at Function.Module._load (module.js:280:25)
    at Module.require (module.js:364:17)
    at require (module.js:380:17)
    at Object.<anonymous> (/home/irritum/stuff/repos/my/bokeh/bokehjs/gulpfile.js:1:63)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Module.require (module.js:364:17)
```
How can I **automatically** install **all** of required dependencies/packages (probably from `package.json` file)?

bryevdv commented:
Run npm install in the bokehjs dir, that should not have changed (it just needs to be done again there are new packages)



> On Apr 15, 2015, at 04:37, Kamil Kwiek (irritum) <notifications@github.com> wrote:
> 
> Hi @bryevdv,
> 
> I'm having some troubles with debugging.
> Right now I'm trying to setup my development environment once again because:
> 
> I was forced to clean up some packages through conda clean subcommand - to gain more free space after 2 years of usage - this breaks a little bit my devbokeh environment,
> recently you have switched from grunt to gulp and made even more changes in code and overall development process (like: #2109),
> documentation on page http://bokeh.pydata.org/en/latest/docs/dev_guide/building.html#building-bokehjs is obsolate.
> Currently I'm getting:
> 
> $ [irritum@localhost bokehjs]$ gulp
> 
> module.js:340
>     throw err;
>           ^
> Error: Cannot find module 'coffee-script/register'
>     at Function.Module._resolveFilename (module.js:338:15)
>     at Function.Module._load (module.js:280:25)
>     at Module.require (module.js:364:17)
>     at require (module.js:380:17)
>     at Object.<anonymous> (/home/irritum/stuff/repos/my/bokeh/bokehjs/gulpfile.js:1:63)
>     at Module._compile (module.js:456:26)
>     at Object.Module._extensions..js (module.js:474:10)
>     at Module.load (module.js:356:32)
>     at Function.Module._load (module.js:312:12)
>     at Module.require (module.js:364:17)
> How can I automatically install all of required dependencies/packages (probably from package.json file)?
> 
> —
> Reply to this email directly or view it on GitHub.
>

mutirri commented:
> Run npm install in the bokehjs dir, that should not have changed.

@bryevdv indeed.

I see that you have rewrite `bokehjs/src/coffee/tool/*` files.
Right now I'm getting my **unique** CSS class **per tool** (inspecting the code in browser through chromium developer tools) - that's very good.
However when I click on the document - nothing happens.
This action should unfold the plot for me.

Below is the single error from `bokeh-server` output (maybe related):
```
2015-04-15 16:20:48,348:WARNING:tornado.access:404 GET /bokehjs/static/js/jsnlog.js.map 127.0.0.1) 102.75ms
```

Of course there is no such file in build directory:
```
[irritum@localhost static]$ find . -iname '*jsnlog.js*'
./js/vendor/jsnlog.js-2.7.5
./js/vendor/jsnlog.js-2.7.5/jsnlog.js
```

bryevdv commented:
@mutirri that warning about the map file is harmless, AFAIK, I see it all the time. I'm not sure what "This action should unfold the plot for me." means tho, can you elaborate?

mutirri commented:
> @mutirri that warning about the map file is harmless, AFAIK, I see it all the time.

OK.

> I'm not sure what "This action should unfold the plot for me." means tho, can you elaborate?

@bryevdv  I'm uploading example to already running bokeh server in `dev` mode:

```
(devbokeh)[irritum@localhost bokeh]$ python bokeh/tests/selenium/examples/simple_line.py 
```

Then I'm opening web browser, navigating to `http://localhost:5006/` site.
I can see one document on the list - my document.
However when I click on it to see the plot - nothing happens, no plot is showing up.
Before the merging master to current pull request this worked just fine.

I even locally revert all of my changes in `bokehjs` directory (re-run the `gulp`) - still no plot.
Are you sure that after last modifications of code the uploading document and showing the plot is still working?

bryevdv commented:
Are the server logs visible? Do they show anything?


> On Apr 16, 2015, at 06:25, Kamil Kwiek (irritum) <notifications@github.com> wrote:
> 
> @mutirri that warning about the map file is harmless, AFAIK, I see it all the time.
> 
> OK.
> 
> I'm not sure what "This action should unfold the plot for me." means tho, can you elaborate?
> 
> @bryevdv I'm uploading example to already running bokeh server in dev mode:
> 
> (devbokeh)[irritum@localhost bokeh]$ python bokeh/tests/selenium/examples/simple_line.py 
> Then I'm opening web browser, navigating to http://localhost:5006/ site.
> I can see one document on the list - my document.
> However when I click on it to see the plot - nothing happens, no plot is showing up.
> Before the merging master to current pull request this worked just fine.
> 
> I even locally revert all of my changes in bokehjs directory (re-run the gulp) - still no plot.
> Are you sure that after last modifications of code the uploading document and showing the plot is still working?
> 
> —
> Reply to this email directly or view it on GitHub.
>

mutirri commented:
As I wrote previously, the only significant error was related to `jsnlog.js.map` file.
Full log:
```
(devbokeh)[irritum@localhost bokeh]$ ./bokeh-server --dev
No handlers could be found for logger "bokeh.resources"

    Bokeh Server Configuration
    ==========================
    python version : 2.7.9
    bokeh version  : 0.9.0dev-c63a78c-61-g17c3946
    listening      : 127.0.0.1:5006
    backend        : memory
    python options : debug:OFF, verbose:OFF, filter-logs:OFF, multi-user:OFF
    js options     : splitjs:OFF, debugjs:ON

/home/irritum/stuff/repos/my/bokeh/bokeh/server/blaze/__init__.py:21: UserWarning: could not import multiuser blaze server No module named mbs.views.  This is fine if you do not intend to use blaze capabilities in the bokeh server
  warnings.warn(msg)
2015-04-16 13:46:49,716:INFO:tornado.access:200 GET /bokeh/userinfo/ (127.0.0.1) 120.58ms
2015-04-16 13:46:51,415:INFO:tornado.access:200 POST /bokeh/doc/ (127.0.0.1) 214.41ms
2015-04-16 13:46:51,509:INFO:tornado.access:200 GET /bokeh/getdocapikey/a5df1718-9649-459b-b7a4-8ccce0447876 (127.0.0.1) 89.81ms
2015-04-16 13:46:52,073:INFO:tornado.access:200 POST /bokeh/bb/a5df1718-9649-459b-b7a4-8ccce0447876/gc (127.0.0.1) 559.99ms
2015-04-16 13:46:52,267:INFO:tornado.access:200 GET /bokeh/bb/a5df1718-9649-459b-b7a4-8ccce0447876/ (127.0.0.1) 80.00ms
2015-04-16 13:47:28,066:INFO:tornado.access:200 POST /bokeh/bb/a5df1718-9649-459b-b7a4-8ccce0447876/bulkupsert (127.0.0.1) 433.09ms
2015-04-16 13:47:45,940:INFO:tornado.access:200 GET / (127.0.0.1) 864.05ms
2015-04-16 13:48:05,633:INFO:tornado.access:200 GET /bokeh/wsurl/ (127.0.0.1) 107.67ms
2015-04-16 13:48:06,342:INFO:tornado.access:200 GET /bokeh/userinfo/ (127.0.0.1) 35.58ms
2015-04-16 13:48:06,598:WARNING:tornado.access:404 GET /favicon.ico (127.0.0.1) 114.36ms
2015-04-16 13:48:45,381:INFO:bokeh.server.views.main:num models: 23
2015-04-16 13:48:45,490:INFO:tornado.access:200 GET /bokeh/bokehinfo/a5df1718-9649-459b-b7a4-8ccce0447876/ (127.0.0.1) 583.89ms
```

Additionally I can see that `gulp` which is running on other console takes almost 90% of CPU resources when I'm doing the click on document name. After a while this load decrease to 15-30%.

bryevdv commented:
@mutirri sorry I am used to seeing that message in a JS console I didn't realize you were taking it from the server logs. I am not sure what to suggest at this point, we definitely have run server examples. I just tried `population_server.py` on master, and it was functioning as expected. @mattpap can you devote a small chunk of time to try and help figure out what is going on here?

mutirri commented:
> @mutirri sorry I am used to seeing that message in a JS console I didn't realize
> you were taking it from the server logs.

I mentioned that this is a `bokeh-server` output :P
Anyway, there is no errors in JS console either:

```
Bokeh: setting prefix to / bokeh.js:40946
Bokeh: setting ws_conn_string to: ws://localhost:5006/bokeh/sub bokeh.js:37484
Bokeh: subscribing to debug bokeh.js:37469
Bokeh: load_models: start bokeh.js:37484
Bokeh: load_models: starting deferred initializations bokeh.js:37484
Bokeh: Canvas initialized bokeh.js:37484
Bokeh: ToolManager: activating tool: PanTool (67db3c1a-3376-4602-abba-d075afacd1e8) for event type 'pan' bokeh.js:37484
Bokeh: ToolManager: activating tool: WheelZoomTool (d74b277f-2038-4309-9ab8-b82737b91a3d) for event type 'scroll' bokeh.js:37484
Bokeh: Plot initialized bokeh.js:37484
Bokeh: load_models: finished deferred initializations bokeh.js:37484
Bokeh: load_models: finish bokeh.js:37484
Bokeh: CanvasView initialized bokeh.js:37484
Bokeh: Line GlyphRenderer (57d8709e-d23f-47c7-8d3f-6e0616830ce9): set_data finished in 27ms bokeh.js:37484
Bokeh: Registering tool: PanTool for event 'pan' bokeh.js:37484
Bokeh: Registering tool: WheelZoomTool for event 'scroll' bokeh.js:37484
Bokeh: Registering tool: BoxZoomTool for event 'pan' bokeh.js:37484
Bokeh: Button tool: PreviewSaveTool bokeh.js:37484
Bokeh: Registering tool: ResizeTool for event 'pan' bokeh.js:37484
Bokeh: Button tool: ResetTool bokeh.js:37484
Bokeh: attaching toolbar to .bk-plot-above for plot da7b327a-d3b0-420f-99a5-c56b007e808a bokeh.js:37484
Bokeh: PlotView initialized bokeh.js:37484
Bokeh: Line GlyphRenderer (57d8709e-d23f-47c7-8d3f-6e0616830ce9): render finished in 6ms bokeh.js:37484
Bokeh: initial ranges set
```

> I am not sure what to suggest at this point, we definitely have run server examples.
> I just tried population_server.py on master, and it was functioning as expected.

Now I tried the most simple example: `examples/plotting/server/line.py` (which generally differs from selenium example by last used option - `show()` instead of `push()`).
And yes, this opens a new tab and the plot is showing up there.

**However** still, when you try to **click** on this document from the main page (`http://localhost:5006/`) - nothing happens (no plot is unfolded as it was before the merge).

I'm not sure if this is related but the URL addresses are also somehow different:
a) URL from main page when I click on the document (no plot): http://localhost:5006/#CF5850C59E0B4C4B89A8E22B8AB1775C
b) URL from new tab in browser (plot exists): http://localhost:5006/bokeh/doc/80bf8786-19b0-4557-bc50-19e8ca121034/367e3bd6-d933-4f31-a06e-400766a8c994

bryevdv commented:
OH It did not click for me that you were referring to the "plot list" page. We have talked about disabling the roll up there for ages, so I am completely unconcerned with that page. It is basically and engineering/debugging page. Clicking a link should open a plot in a new tab, but I don't care about that either, for this PR. (Please make a separate  issue if clicking does not open a tab with a plot). We should probably have had a ghangout or other higher-bandwidth communication about this earlier on. Just to be clear, for this PR, I *only* care about testing actual plot pages. It sounds like you can navigate to actual plot pages now, correct?

mutirri commented:
Hi @bryevdv,

> OH It did not click for me that you were referring to the "plot list" page.
> We have talked about disabling the roll up there for ages,
> so I am completely unconcerned with that page.
> It is basically and engineering/debugging page.

I didn't know that.
This was working before the merge and now is broken.

> Clicking a link should open a plot in a new tab, but I don't care about that either, for this PR.

It's not working.
I was using this page to check uploaded examples.

> (Please make a separate issue if clicking does not open a tab with a plot).

Done: https://github.com/bokeh/bokeh/issues/2187

> We should probably have had a ghangout or other higher-bandwidth communication
> about this earlier on.
 
Yeah :)

> Just to be clear, for this PR, I only care about testing actual plot pages.
> It sounds like you can navigate to actual plot pages now, correct?

Indeed.
I rewrote the mechanism and now I'm able to navigate directly to uploaded examples.

mutirri commented:
Hi @bryevdv,

Just FYI.

1. Currently I finished tests for such tools: `box_zoom`, `pan`, `reset`, `resize`.
A) Under `Firefox` driver **all** the tests are working perfectly **fine**.
B) Under `PhantomJS` (due to limitations of this engine) I'm getting unexpected results in some tests.

2. I'm having also some issues with writting appropriate test for `wheel_zoom` tool.
That's because there is no native support for such action in selenium framework and 
I'm forced to invent and directly inject appropriate JS instead of.

bryevdv commented:
@mutirri That is good news! I would say at this point, the most important next step is to gain some basic coverage widgets/server examples. I would not worry about `wheel_zoom` at all at this point.

mutirri commented:
Hi,

> I would say at this point, the most important next step is to gain some
> basic coverage widgets/server examples.

Sure, no problem.
However be patient because I'm switching to bokeh between my other tasks.

> I would not worry about wheel_zoom at all at this point.

OK.

Best.

bryevdv commented:
@mutirri I understand. Please feel free to ping @mattpap with Bokeh specific questions as well.

damianavila commented:
@mutirri do you hace any update on this? Thanks!

mutirri commented:
Hi @damianavila,

> do you hace any update on this?

Sorry but no.
I'm still engaged in other tasks.
Moreover in the next week I have two days off also.

damianavila commented:
OK, @mutirri, thanks for the report!!
@bryevdv, I don't think we should keep this PR open anymore... @mutirri can work in a branch an reopen it again when it is ready for review. Thoughts?

bryevdv commented:
Yes, let's close it. @birdsarah has Selenium experience and may have more availability to help stand something at some point. She can refer to the closed PR as a reference to what has already been done.

birdsarah commented:
looking forward to getting my teeth into this one.
